File: app/src/test/java/jp/toastkid/yobidashi/browser/tab/TabTest.java
package jp.toastkid.yobidashi.browser.tab;

import android.support.annotation.NonNull;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

import org.junit.Test;

import java.io.IOException;

import static org.junit.Assert.assertEquals;

/**
 * {@link Tab}'s test case.
 *
 * @author toastkidjp
 */
public class TabTest {

    @Test
    public void test() throws IOException {
        final Tab tab = makeTestTab();

        final JsonAdapter<Tab> tabJsonAdapter = makeTabJsonAdapter();
        final String json = check_toJson(tab, tabJsonAdapter);

        check_fromJson(tabJsonAdapter, json);
    }

    private void check_fromJson(JsonAdapter<Tab> tabJsonAdapter, String json) throws IOException {
        final Tab fromJson = tabJsonAdapter.fromJson(json);
        assertEquals("Google",    fromJson.getLastTitle());
        assertEquals("file://~~", fromJson.getThumbnailPath());
        assertEquals("Title",     fromJson.getLatest().title());
        assertEquals("URL",       fromJson.getLatest().url());
    }

    private String check_toJson(Tab tab, JsonAdapter<Tab> tabJsonAdapter) {
        final String json = tabJsonAdapter.toJson(tab);
        assertEquals(
                "{\"histories\":[{\"title\":\"Title\",\"url\":\"URL\"}]," +
                        "\"index\":0,\"lastTitle\":\"Google\",\"thumbnailPath\":\"file://~~\"}",
                json
        );
        return json;
    }

    private JsonAdapter<Tab> makeTabJsonAdapter() {
        final Moshi moshi = new Moshi.Builder().build();
        return moshi.adapter(Tab.class);
    }

    @NonNull
    private Tab makeTestTab() {
        final Tab tab = new Tab();
        tab.setThumbnailPath("file://~~");
        tab.setLastTitle("Google");
        tab.addHistory(History.Companion.make("Title", "URL"));
        return tab;
    }
}

File: app/src/test/java/jp/toastkid/yobidashi/browser/tab/TabTest.kt
package jp.toastkid.yobidashi.browser.tab

import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.IOException

/**
 * [Tab]'s test case.
 *
 * @author toastkidjp
 */
@RunWith(RobolectricTestRunner::class)
class TabTest {

    @Test
    @Throws(IOException::class)
    fun test() {
        val tab = makeTestTab()

        val tabJsonAdapter = makeTabJsonAdapter()
        val json = check_toJson(tab, tabJsonAdapter)

        check_fromJson(tabJsonAdapter, json)
    }

    @Throws(IOException::class)
    private fun check_fromJson(tabJsonAdapter: JsonAdapter<Tab>, json: String) {
        val fromJson = tabJsonAdapter.fromJson(json)
        assertEquals("file://~~", fromJson?.thumbnailPath)
        assertEquals("Title", fromJson?.latest?.title())
        assertEquals("URL", fromJson?.latest?.url())
    }

    private fun check_toJson(tab: Tab, tabJsonAdapter: JsonAdapter<Tab>): String {
        val json = tabJsonAdapter.toJson(tab)
        assertTrue(json.contains("\"histories\":[{\"title\":\"Title\",\"url\":\"URL\"}]"))
        assertTrue(json.contains("\"index\":0"))
        assertTrue(json.contains("\"thumbnailPath\":\"file://~~\""))
        return json
    }

    private fun makeTabJsonAdapter(): JsonAdapter<Tab> {
        val moshi = Moshi.Builder().build()
        return moshi.adapter(Tab::class.java)
    }

    private fun makeTestTab(): Tab {
        val tab = Tab()
        tab.thumbnailPath = "file://~~"
        tab.addHistory(History.make("Title", "URL"))
        return tab
    }
}
